text/src/autogluon/text/automm/optimization/lit_distiller.py

```diff
@@ -0,0 +1,384 @@
+        self.soft_label_loss_func = soft_label_loss_func
+        if isinstance(validation_metric, BaseAggregator) and custom_metric_func is None:
+            raise ValueError(
+                f"validation_metric {validation_metric} is an aggregation metric,"
+                f"which must be used with a customized metric function."
+            )
+        self.custom_metric_func = custom_metric_func
+
+    def _compute_hard_label_loss(
+            self,
+            output: dict,
+            label: torch.Tensor,
+    ):
+        loss = 0
+        for _, per_output in output.items():
+            weight = per_output[WEIGHT] if WEIGHT in per_output else 1
+            loss += self.hard_label_loss_func(
+                input=per_output[LOGITS].squeeze(dim=1),
+                target=label,
+            ) * weight
+        return loss
+
+    def _compute_soft_label_loss(
+            self,
+            student_output: dict,
+            teacher_output: dict,
+    ):
+        student_logits = student_output[self.student_model.prefix][LOGITS].squeeze(dim=1)
+        soft_labels = teacher_output[self.teacher_model.prefix][LOGITS].squeeze(dim=1)
+        student_logits = student_logits / self.temperature
+        soft_labels = soft_labels / self.temperature
+
+        if isinstance(self.soft_label_loss_func, nn.CrossEntropyLoss):
+            soft_labels = F.softmax(soft_labels, dim=-1)
+
+        loss = self.soft_label_loss_func(
+            input=student_logits,
+            target=soft_labels,
+        )
+        return loss
+
+    def _compute_loss(
+            self,
+            student_output: dict,
+            teacher_output: dict,
+            label: torch.Tensor,
+    ):
+        loss = 0
+        hard_label_loss = self._compute_hard_label_loss(
+            output=student_output,
+            label=label,
+        )
+        loss += hard_label_loss * self.hard_label_weight
+
+        soft_label_loss = self._compute_soft_label_loss(
+            student_output=student_output,
+            teacher_output=teacher_output,
+        )
+        loss += soft_label_loss * self.soft_label_weight
+
+        return loss
+
+    def _compute_metric(
+            self,
+            logits: torch.Tensor,
+            label: torch.Tensor,
+    ):
+        if isinstance(self.validation_metric, torchmetrics.AUROC):
+            prob = F.softmax(logits.float(), dim=1)
+            return self.validation_metric(preds=prob[:, 1], target=label)  # only for binary classification
+        elif isinstance(self.validation_metric, BaseAggregator):
+            return self.validation_metric(self.custom_metric_func(logits, label))
+        else:
+            return self.validation_metric(logits.squeeze(dim=1), label)
+
+    def _shared_step(
+            self,
+            batch: dict,
+    ):
+        student_output = self.student_model(batch)
+        self.teacher_model.eval()
+        with torch.no_grad():
+            teacher_output = self.teacher_model(batch)
```

Comment:
Does it makes sense to cache teacher outputs? If we do this more than one epoch, this would save compute.

Reply:
This is a trade-off between time and space. Caching teacher outputs can increase memory usage, which could be an issue for large datasets. In future, we will add distillation methods using hidden features. Caching the hidden features would have more memory cost.